### PR TITLE
Adding unit tests for CacheBackedIdPMgtDAO.java

### PR DESCRIPTION
**Pull request by ZiyamSanthosh**
### Proposed changes in this pull request

- A new java class named CacheBackedIdPMgtDAOTest.java containing unit tests written for public methods in CacheBackedIdPMgtDAO.java
- Above mentioned new test java class is added to relevant testng.xml file

### CODE REVIEW

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java [hunk 1]
```diff
@@ -0,0 +1,1460 @@
...
+import org.apache.commons.lang.StringUtils;
+import org.powermock.core.classloader.annotations.PrepareForTest;
+import org.powermock.modules.testng.PowerMockTestCase;
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+import org.wso2.carbon.base.MultitenantConstants;
+import org.wso2.carbon.context.CarbonContext;
+import org.wso2.carbon.identity.application.common.model.*;
```
Comment (ZiyamSanthosh) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java:31]: Remove bulk imports

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java [hunk 2]
```diff
@@ -0,0 +1,1460 @@
...
+
+        try (Connection connection = getConnection(DB_NAME)) {
+            when(IdentityDatabaseUtil.getDBConnection(anyBoolean())).thenReturn(connection);
+            when(IdentityDatabaseUtil.getDataSource()).thenReturn(dataSourceMap.get(DB_NAME));
+            addTestIdps();
+
+            IdentityProvider idpResult = cacheBackedIdPMgtDAO.getIdPByName(connection, idpName, tenantId, TENANT_DOMAIN);
+
+            IdentityProvider idpResult2 = new IdentityProvider();
+            if (isExist){
```
Comment (ZiyamSanthosh) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java:363]: space between brackets

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java [hunk 3]
```diff
@@ -0,0 +1,1518 @@
...
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.mockito.Matchers.anyBoolean;
+import static org.mockito.Matchers.anyInt;
+import static org.powermock.api.mockito.PowerMockito.mock;
```
Comment (nilasini) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java:75]: try to not to use powermockito

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java [hunk 4]
```diff
@@ -0,0 +1,1518 @@
...
+
+        closeH2Database();
+    }
+
+    @DataProvider
+    public Object[][] getIdPsData() {
+
+        return new Object[][]{
+                {SAMPLE_TENANT_ID, TENANT_DOMAIN, 2},
+                {SAMPLE_TENANT_ID2, TENANT_DOMAIN, 1},
```
Comment (nilasini) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java:162]: add description of each arguments

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java [hunk 5]
```diff
@@ -0,0 +1,1518 @@
...
+                {4, TENANT_DOMAIN, 0}
```
Comment (nilasini) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java:163]: why some tenant ids added as constant and only one is not added as constant

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java [hunk 6]
```diff
@@ -0,0 +1,1518 @@
...
+        };
+    }
+
+    @Test(dataProvider = "getIdPsData")
+    public void testGetIdPs(int tenantId, String tenantDomain, int idpCount) throws Exception {
+
+        mockStatic(IdentityDatabaseUtil.class);
+
+        try (Connection connection = getConnection(DB_NAME)) {
+
```
Comment (nilasini) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java:173]: remove extra line

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java [hunk 7]
```diff
@@ -0,0 +1,1518 @@
...
+            when(IdentityDatabaseUtil.getDBConnection(anyBoolean())).thenReturn(connection);
+            when(IdentityDatabaseUtil.getDataSource()).thenReturn(dataSourceMap.get(DB_NAME));
+            addTestIdps();
+
+            List<IdentityProvider> idps1 = cacheBackedIdPMgtDAO.getIdPs(connection, tenantId, tenantDomain);
```
Comment (nilasini) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java:178]: Change variable name to more meaning full one. Suggestion idpList1

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java [hunk 8]
```diff
@@ -0,0 +1,1518 @@
...
+            assertEquals(idps1.size(), idpCount);
+            List<IdentityProvider> idps2 = cacheBackedIdPMgtDAO.getIdPs(null, tenantId, tenantDomain);
```
Comment (nilasini) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java:180]: Add comment to understand the difference between first getIdps call and second call

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java [hunk 9]
```diff
@@ -0,0 +1,1518 @@
...
+        mockStatic(IdentityDatabaseUtil.class);
+
+        try (Connection connection = getConnection(DB_NAME)) {
+            when(IdentityDatabaseUtil.getDBConnection(anyBoolean())).thenReturn(connection);
+            when(IdentityDatabaseUtil.getDBConnection()).thenReturn(connection);
+            when(IdentityDatabaseUtil.getDataSource()).thenReturn(dataSourceMap.get(DB_NAME));
+            addTestIdps();
+
+            List<IdentityProvider> idps1 = cacheBackedIdPMgtDAO.getIdPsSearch(connection, tenantId,
+                    tenantDomain, filter);
```
Comment (nilasini) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java:227]: formatting issue

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java [hunk 10]
```diff
@@ -0,0 +1,1518 @@
...
+
+        mockStatic(IdentityDatabaseUtil.class);
+
+        try (Connection connection = getConnection(DB_NAME)) {
+            when(IdentityDatabaseUtil.getDBConnection(anyBoolean())).thenReturn(connection);
+            when(IdentityDatabaseUtil.getDBConnection()).thenReturn(connection);
+            when(IdentityDatabaseUtil.getDataSource()).thenReturn(dataSourceMap.get(DB_NAME));
+            addTestIdps();
+
+            List<IdentityProvider> idps1 = cacheBackedIdPMgtDAO.getIdPsSearch(connection, tenantId,
```
Comment (nilasini) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java:226]: change variable name

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java (hunk 3 above)
Comment (ZiyamSanthosh) [reply components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java:75]: The version of mockito used is old which doesn't have the mockstatic which is available in PowerMockito. Updating mockito to its latest version breaks other test cases

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java (hunk 5 above)
Comment (ZiyamSanthosh) [reply components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java]: Introduced constants

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java (hunk 7 above)
Comment (ZiyamSanthosh) [reply components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java]: Meaningful variable names added

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java (hunk 8 above)
Comment (ZiyamSanthosh) [reply components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java]: Added comments

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java (hunk 10 above)
Comment (ZiyamSanthosh) [reply components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java]: Meaningful variable names added

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java [hunk 11]
```diff
@@ -0,0 +1,1486 @@
...
+        IdPNameCacheKey cacheKey = new IdPNameCacheKey(idpName);
+        IdPCacheEntry entry = idPCacheByName.getValueFromCache(cacheKey, TENANT_DOMAIN);
+        if (entry != null) {
+            return entry.getIdentityProvider();
+        } else {
+            return null;
+        }
+    }
+
+    private IdentityProvider idpFromCacheByResourceId(String resourceId) {
```
Comment (nilasini) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java:1476]: Add new line after the method. Please check all places

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java [hunk 12]
```diff
@@ -0,0 +1,1486 @@
...
+/**
+ * Unit tests for CacheBackedIdPManagementDAO.
+ */
+@PrepareForTest({IdentityDatabaseUtil.class, DataSource.class, IdentityTenantUtil.class, IdentityUtil.class,
+        CarbonContext.class})
+public class CacheBackedIdPMgtDAOTest extends PowerMockTestCase {
+
+    private static final String DB_NAME = "test";
+
+    private static final Integer SAMPLE_TENANT_ID = -1234;
```
Comment (nilasini) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java:92]: Let's use SAMPLE_TENANT_ID1

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java [hunk 13]
```diff
@@ -0,0 +1,1486 @@
...
+
+        closeH2Database();
+    }
+
+    @DataProvider
+    public Object[][] getIdPsData() {
+
+        return new Object[][]{
+                {SAMPLE_TENANT_ID, TENANT_DOMAIN, 2},
+                {SAMPLE_TENANT_ID2, TENANT_DOMAIN, 1},
```
Comment (nilasini) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java:164]: All the case we are passing TENANT_DOMAIN there is not change in all 3 combination then no need to pass them right?

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java [hunk 14]
```diff
@@ -0,0 +1,1486 @@
...
+                cacheBackedIdPMgtDAO.getIdPs(null, tenantId, tenantDomain));
+    }
+
+    @DataProvider
+    public Object[][] getIdPsSearchData() {
+
+        return new Object[][]{
+                {SAMPLE_TENANT_ID, TENANT_DOMAIN, "testIdP1", 1},
+                {SAMPLE_TENANT_ID2, TENANT_DOMAIN, "testIdP3", 1},
+                {SAMPLE_TENANT_ID, TENANT_DOMAIN, "", 2},
```
Comment (nilasini) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java:207]: All scenarios getting same tenant domain? Shall we use different tenant domain combination as well?

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java [hunk 15]
```diff
@@ -0,0 +1,1486 @@
...
+        ExpressionNode expressionNode = new ExpressionNode();
+        List<ExpressionNode> expressionNodeList = new ArrayList<>();
+        expressionNodeList.add(expressionNode);
+
+        List<String> attributesList1 = Arrays.asList("id", "name", "description", "isEnabled", "image", "isPrimary");
+        List<String> attributesList2 = Arrays.asList("homeRealmIdentifier", "isFederationHub", "certificate", "alias",
+                "claims", "roles", "federatedAuthenticators", "provisioning");
+
+        return new Object[][]{
+                {SAMPLE_TENANT_ID, expressionNodeList, 2, 0, "ASC", "NAME", attributesList1, 2},
```
Comment (nilasini) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java:265]: Same value passed for all combination as SAMPLE_TENANT_ID

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java [hunk 16]
```diff
@@ -0,0 +1,1486 @@
...
+    @DataProvider
+    public Object[][] getPaginatedIdPsSearchExceptionData() {
+
+        ExpressionNode expressionNode = new ExpressionNode();
+        List<ExpressionNode> expressionNodes = new ArrayList<>();
+        expressionNodes.add(expressionNode);
+        List<String> attributes = Arrays.asList("WrongAttribute");
+
+        return new Object[][]{
+                {SAMPLE_TENANT_ID, expressionNodes, 2, 0, "ASC", "NAME", attributes}
```
Comment (nilasini) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java:296]: if we don't checking multiple combination then no need to create a data provider.

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java [hunk 17]
```diff
@@ -0,0 +1,1486 @@
...
+
+    @Test(dataProvider = "getIdPByNameData")
+    public void testGetIdPByName(String idpName, int tenantId, boolean isExist) throws Exception {
+
+        mockStatic(IdentityDatabaseUtil.class);
+        try (Connection connection = getConnection(DB_NAME)) {
+            when(IdentityDatabaseUtil.getDBConnection(anyBoolean())).thenReturn(connection);
+            when(IdentityDatabaseUtil.getDataSource()).thenReturn(dataSourceMap.get(DB_NAME));
+            addTestIdps();
+            //retrieving IDP from DB and adding to cache
```
Comment (nilasini) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java:382]: Add a space before start the comment, All comments should start with capital and ends with full stop.

Comment (nilasini) [reply components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java:382]: Check other places as well

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java [hunk 18]
```diff
@@ -0,0 +1,1486 @@
...
+        }
+    }
+
+    @DataProvider
+    public Object[][] getIdPByAuthenticatorPropertyValueWithoutAuthenticatorData() {
+
+        return new Object[][]{
+                {SAMPLE_TENANT_ID, "testIdP1", "Property1", "value1", true},
+                {SAMPLE_TENANT_ID, "testIdP1", "Property2", "value2", true},
+                {SAMPLE_TENANT_ID, "NotExist", "Null", "Null", false},
```
Comment (nilasini) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java:535]: Same value for SAMPLE_TENANT_ID passed in all combination

Comment (nilasini) [reply components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java:535]: Check other places as well

**Review by nilasini — APPROVED** [review]

**Comment by senthalan** [comment]
Closing as the PR is in the open state for a while. Please feel free to reopen and merge if needed.

**Comment by nilasini** [comment]
Only one intermittent test is failing hence merging this PR
"UserMeApprovalTest.testListTasksWhenAvailable:141 1 expectation failed."
https://github.com/wso2/product-is/actions/runs/2802685968

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java (hunk 11 above)
Comment (ZiyamSanthosh) [reply components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java:1469]: Resolved with https://github.com/wso2/carbon-identity-framework/pull/4110/commits/28a1194ed9f5de659bd1879f625cd070efdac3fa

Comment (ZiyamSanthosh) [reply components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java:1469]: Resolved with https://github.com/wso2/carbon-identity-framework/pull/4110/commits/28a1194ed9f5de659bd1879f625cd070efdac3fa

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java (hunk 12 above)
Comment (ZiyamSanthosh) [reply components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java]: Resolved with https://github.com/wso2/carbon-identity-framework/pull/4110/commits/28a1194ed9f5de659bd1879f625cd070efdac3fa

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java (hunk 13 above)
Comment (ZiyamSanthosh) [reply components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java]: Resolved with https://github.com/wso2/carbon-identity-framework/pull/4110/commits/28a1194ed9f5de659bd1879f625cd070efdac3fa

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java (hunk 14 above)
Comment (ZiyamSanthosh) [reply components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java]: Resolved with https://github.com/wso2/carbon-identity-framework/pull/4110/commits/28a1194ed9f5de659bd1879f625cd070efdac3fa

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java (hunk 15 above)
Comment (ZiyamSanthosh) [reply components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java]: Resolved with https://github.com/wso2/carbon-identity-framework/pull/4110/commits/28a1194ed9f5de659bd1879f625cd070efdac3fa

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java (hunk 17 above)
Comment (ZiyamSanthosh) [reply components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java]: Resolved with https://github.com/wso2/carbon-identity-framework/pull/4110/commits/28a1194ed9f5de659bd1879f625cd070efdac3fa

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java (hunk 18 above)
Comment (ZiyamSanthosh) [reply components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java]: Resolved with https://github.com/wso2/carbon-identity-framework/pull/4110/commits/28a1194ed9f5de659bd1879f625cd070efdac3fa

**Pull request merged**